### DISCUSSION TITLE
[v2] Collections do not work well with phpstan

### DISCUSSION THREAD

**Discussion by robertvansteen** [spatie/laravel-data#158] [Q&A]
Thanks for releasing v2, looks very promising. One thing I did notice is that the changes to collections do not seem to play well with phpstan.

If you have something like
```php
class Payload extends Data
{
    public function __construct(
        /** @var DataCollection<User> */
        public DataCollection $users,
    ) {}
}
```
```php
new Payload(
   users: UserData::collection(User::all())
);
```
it will give the following phpstan error
```
Parameter $users of class App\Payload constructor iterable<App\UserData>&Spatie\LaravelData\DataCollection,                             
       Spatie\LaravelData\CursorPaginatedDataCollection|Spatie\LaravelData\DataCollection|Spatie\LaravelData\PaginatedDataCollection given. 
```
So it seems that phpstan is complaining about the fact that calling `Data::collection` could return either a normal collection, a paginated or a cursor paginated data collection. For strongly typed codebases it would probably be helpful to add methods that explicitly returns one of those three alongside the `collection` static method that could return any of those three.

Is this something you'd consider if I would to open a PR for? Or is there any other way to satisfy static analysers?

**Comment by robertvansteen** [comment]
@rubenvanassche any thoughts about this?

**Reply by rubenvanassche** [reply]
What names would you give the methods? `paginatedCollection` and `cursorPaginatedCollection` seem clear to me but what about the `collection` method?

**Reply by robertvansteen** [reply]
That's tricky yeah. I decided to look into conditional return types for phpstan and opened a PR with a different solution to this: #184

**Reply by iamrgroot** [reply]
I commented on #184, but as that is already merged I'll ask my question here as well:

Would it be possible to add the TKey and TValue in the return types of the collection function?

Something like `\Spatie\LaravelData\DataCollection<int,static>` instead of `\Spatie\LaravelData\DataCollection`?

This prevents me from having to do it myself in my Data objects to make phpstan happy
